File: web/src/crypto/transcript.ts
/**
 * Length-prefixed concatenation, per protocol.md §3.3: every field is
 * preceded by its length as a uint16 big-endian, so no two distinct field
 * splits can ever produce the same transcript bytes (unlike naive
 * concatenation, where e.g. "ab"+"c" and "a"+"bc" collide).
 */
export function buildTranscript(fields: Uint8Array[]): Uint8Array {
  let total = 0
  for (const f of fields) {
    if (f.length > 0xffff) throw new Error('transcript field exceeds uint16 length prefix')
    total += 2 + f.length
  }

  const out = new Uint8Array(total)
  let offset = 0
  for (const f of fields) {
    out[offset] = (f.length >> 8) & 0xff
    out[offset + 1] = f.length & 0xff
    offset += 2
    out.set(f, offset)
    offset += f.length
  }
  return out
}

/**
 * ASCII-only encoder used for protocol constants and base64 strings (the
 * only inputs this function is ever called with). Deliberately avoids the
 * global TextEncoder: under some test environments (jsdom) it returns a
 * Uint8Array from a different realm than `new Uint8Array()`, which
 * libsodium's strict type checks reject.
 */
export function utf8(s: string): Uint8Array {
  const out = new Uint8Array(s.length)
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i)
    if (code > 0x7f) throw new Error('utf8() only supports ASCII input')
    out[i] = code
  }
  return out
}

export function u8(n: number): Uint8Array {
  return Uint8Array.of(n & 0xff)
}

File: android/app/src/main/java/com/depot/app/crypto/Transcript.kt
package com.depot.app.crypto

import java.io.ByteArrayOutputStream

/**
 * Length-prefixed concatenation, per protocol.md §3.3: every field is
 * preceded by its length as a uint16 big-endian, so no two distinct field
 * splits can produce the same bytes (unlike naive concatenation, where
 * "ab"+"c" and "a"+"bc" collide).
 */
fun buildTranscript(fields: List<ByteArray>): ByteArray {
    val out = ByteArrayOutputStream()
    for (f in fields) {
        require(f.size <= 0xffff) { "transcript field exceeds uint16 length prefix" }
        out.write((f.size ushr 8) and 0xff)
        out.write(f.size and 0xff)
        out.write(f)
    }
    return out.toByteArray()
}

fun u8(n: Int): ByteArray = byteArrayOf((n and 0xff).toByte())

fun u32be(n: Int): ByteArray = byteArrayOf(
    ((n ushr 24) and 0xff).toByte(),
    ((n ushr 16) and 0xff).toByte(),
    ((n ushr 8) and 0xff).toByte(),
    (n and 0xff).toByte(),
)

fun u64be(n: Long): ByteArray {
    val out = ByteArray(8)
    for (i in 0 until 8) out[i] = ((n ushr (56 - i * 8)) and 0xff).toByte()
    return out
}

fun readU32be(bytes: ByteArray, offset: Int): Int =
    ((bytes[offset].toInt() and 0xff) shl 24) or
        ((bytes[offset + 1].toInt() and 0xff) shl 16) or
        ((bytes[offset + 2].toInt() and 0xff) shl 8) or
        (bytes[offset + 3].toInt() and 0xff)

fun readU64be(bytes: ByteArray, offset: Int): Long {
    var v = 0L
    for (i in 0 until 8) v = (v shl 8) or (bytes[offset + i].toLong() and 0xff)
    return v
}

/**
 * The web side uses an ASCII-only encoder here to sidestep a jsdom realm
 * quirk; for the protocol constants and base64 strings that reach this
 * function, UTF-8 produces identical bytes.
 */
fun utf8(s: String): ByteArray = s.toByteArray(Charsets.UTF_8)
